keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1237 @@
+
+**Note: Option 1 is a *preferred* design option as it is easier to implement, and also easier to understand from UX perspective. 
+If there are enough supporting use cases for Option 2, we can explore the implementation in the beta phase of the KEP. Option 2
+will also require discussions about dynamic QoS classes as per the execution phase of the pod (init vs non-init phase).**
+
+
+#### Admission Controller
+
+* LimitRanger
+
+  LimitRanger today supports objects of type `Container`, `Pod` and `PersistentVolumeClaim`. It allows following constraints:
+  
+    * min (this option is to specify min usage constraint on the resource)
+    * max (this option is to specify max usage constraints on the resource)
+    * defaultRequest (this option is to specify default resource request for the named resource)
+    * default (this option is to specify default limit for the named resource)
+    * maxLimitRequestRatio (represents the max burst value of the named resource)
+  
+  default and defaultRequest are not supported by LimitRanger of type `Pod`.
+
+  To incorporate Pod level resource specifications, LimitRanger logic needs to be modfied for Pod and Container types. For PersistentVolumeClaim, 
+  LimitRanger will remain unchanged as PVCs are independent resources and their size is not directly related to the resources consumed by a pod. 
+  Following are the two design options to consider:
+  
+  * **Option 1: Enhance existing `Pod` type**
+      
+      * Add support for default pod request (`defaultRequest`) and limit (`default`) in the `Pod` LimitRange object to allow customers
+      to specify default requests and limits for pods in a namespace. This requires changing the 
+      (validation logic)[https://github.com/kubernetes/kubernetes/blob/9e2075b3c87061d25759b0ad112266f03601afd8/pkg/apis/core/validation/validation.go#L6403]
+      to allow `defaultRequest` and `default` in Pod LimitRange object.
+
+      * Conditional logic
+        
+         * Pod level resource spec is set and `Pod` LimitRanger is applied.
+           If pod level resources are specified, use these values directly against the Pod type LimitRanger. It doesn't matter if container level
+           resource spec is set or not.
+           
+         * Pod level resource spec is set and `Pod` LimitRanger is applied.
+           If pod level resources are not specified, there could be cases:
+             1. Existing workloads that have `Pod` LimitRanger set to validate the constraints at container level aggregates (and not at pod level).
+             2. New workloads that want to use Pod Level resource limits feature and have set `Pod` LimitRanger to validate constraints at Pod Level.
+           
+           If the behavior of `Pod` LimitRanger is modified, it could break existing workloads and also remove the ability for usecase where users 
```

Comment:
I am unclear what breaks here?  It seems like this is the OBVIOUS extension, no?  

We already have an "effective" pod resources, as detailed above.  What does `LimitRange` do for `type: Pod` today -- I would assume (bad, I know) that it checks against that effective value, am I wrong?

Reply:
LimitRange of type "Pod" already exists today.

LimitRange of type Pod sums up the resources specified for containers as effective values for pod. It is easy to change this to use pod level values with this KEP. But the problem is, today it doesn't support specifying defaultRequest, and default (which is default limit) at pod level since there's no way (without this KEP) to specify these values at Pod level.

Now with the support for defaultRequest and limit in this KEP, these values will be applied when pod level resources are not specified. How to differentiate between when workloads do not specify pod level values so that they can use default values from limit ranger vs when pod level values are not specified for workloads that want to use container level limits only (like in existing workloads)? I might have overcomplicated this, but I wanted to discuss this scenario.

On the contrary, we could say that since defaultRequest and limit can be specified only after pod level resources feature is supported (since today there is no way to specify pod level resources and hence defaultRequest and limits are not supported without this feature), it essentially means that user wants to apply defaultRequest and limit at pod level, if specified in the LimitRange object of type Pod, for all workloads in that namespace including existing workloads. Hence we could leave it to the users to make sure they set defaultRequest and limit only in those namespaces in which they want to support these values as pod level.